---
output:
  pdf_document: default
  html_document: default
---
```{r}
library(HardyWeinberg)
library(data.table)
genotype_data = fread("TSIChr22v4.raw")
df <- data.frame(genotype_data)
df <- df[, (7:ncol(df))]; head(df)
```

# 1. How many variants are there in this database? What percentage of the data is missing?
```{r}
n_variants = ncol(df)
cat("Number of variants in the dataset:", n_variants)

# Convert non-0, 1, or 2 values to NA
df[!sapply(df, function(x) x %in% c(0, 1, 2))] <- NA

n_missing = sum(is.na(df))
total_data_points = ncol(df) * nrow(df)
perc_missing = (n_missing / total_data_points)*100
cat("\nPercentages missing data:", perc_missing)
```
# 2. Calculate the percentage of monomorphic variants. Exclude all monomorphics from the database for all posterior computations of the practical. How many variants do remain in your database?
```{r}
monomorphic <- sapply(df, function(x) length(unique(na.omit(x))) == 1)
cat("Percentage of monomorphic variants:", (sum(monomorphic)/n_variants)*100, "%\n")
df_mono <- df[, !monomorphic]
cat(ncol(df_mono), "variants remain in the database")
```

# 3. Extract polymorphism rs587756191 T from the data, and determine its genotype counts. Apply a chi-square test for Hardy-Weinberg equilibrium, with and without continuity correction. Also try an exact test, and a permutation test. You can use thee functions HWChisq, HWExact and HWPerm for this purpose. Do you think this variant is in equilibrium?

```{r}
specific_variant <- df_mono[,"rs587756191_T"] 
genotype_counts <- c(
  AA=sum(specific_variant==0),
  AB=sum(specific_variant==1),
  BB=sum(specific_variant==2)
)

# Chi-square test without continuity correction
hw_chisq <- HWChisq(genotype_counts, cc=0, verbose = TRUE)

# Chi-square test with continuity correction
hw_chisq_corrected <- HWChisq(genotype_counts, cc = 0.5, verbose = TRUE)

# Exact test for Hardy-Weinberg Equilibrium
hw_exact <- HWExact(genotype_counts, verbose = TRUE)

# Permutation test for Hardy-Weinberg Equilibrium
hw_perm <- HWPerm(genotype_counts, nperm = 10000, verbose = TRUE)

```
# 4. Determine the genotype counts for all polymorphic variants, and store them in a p × 3 matrix.
```{r}
p <- ncol(df_mono)

# Create a matrix to store genotype counts
genotype_counts_all <- matrix(0, nrow = p, ncol = 3)

for (i in 1:p) {
  # Compute genotype couns
  genotype_counts_all[i, 1] <- sum(df_mono[, i] == 0)  # Homozygous for allele 1 AA
  genotype_counts_all[i, 2] <- sum(df_mono[, i] == 1)  # Homozygous for allele 2 AB
  genotype_counts_all[i, 3] <- sum(df_mono[, i] == 2)  # Heterozygous BB
}

print(genotype_counts_all)
```

# 5. Apply an exact test for Hardy-Weinberg equilibrium to each SNP. You can use function HWExactStats for fast computation. What is the percentage of significant SNPs (use α = 0.05)? Is this the number of markers that you would expect to be out of equilibrium by the effect of chance alone?
```{r}
# compute exact test for each variant
alpha <- 0.05
hw_exact_all <- HWExactStats(genotype_counts_all)

# Calculate percentage of variants that are significant with alpha = 0.05
count_below_005 <- sum(hw_exact_all < alpha)
perc_below_005 <- (count_below_005 / length(hw_exact_all)) * 100
cat("Percentage of significant SNPs:", perc_below_005, "%\n")
```
We find that when using Fisher's exact test to test for HWE, the percentage of significant SNPs is 2.77%, and thus for these variants we will reject the null hypothesis that they are in HWE with a significance level of $0.05$. This is the proportion of the variants we would expect to deviate from HWE due to pure chance alone. So if we investigate for HWE with a significance level of $0.05$, we would want the percentage of significant variants to be below $0.05$, thus indicating that there are other factors influencing the deviations from the HWE in the variants.


# 6. Which SNP is most significant according to the exact test results? Give its genotype counts. In which sense is this genotypic composition unusual?
```{r}
min_idx <- which.min(hw_exact_all)
most_significant_variant_name <- names(df_mono)[min_idx]
cat("Most significant variant is:",most_significant_variant_name, "with a p-value of ", min(hw_exact_all))

most_significant_variant_values <- df_mono[,most_significant_variant_name] 

# Get genotype counts
genotype_counts <- c(
  sum(most_significant_variant_values==0),
  sum(most_significant_variant_values==1),
  sum(most_significant_variant_values==2)
)

cat("\nGenotype counts:\nAA:", genotype_counts[1], "\nAB:", genotype_counts[2], "\nBB:", genotype_counts[3])

observed_frequencies <- genotype_counts / nrow(df_mono)

p <- ((2*genotype_counts[1])+(genotype_counts[2]))/(2*nrow(df_mono)) # observed allele frequency for A
q <- 1 - p # observed allele frequency for B

# Expected genotype frequencies under HWE:
exp_AA_freq <- p^2
exp_AB_freq <- 2*p*q
exp_BB_freq <- q^2


cat("\nHWE expected AA freqeuncy:", exp_AA_freq, "vs. observed frequency:", observed_frequencies[1],
    "\nHWE expected AB frequency: ", exp_AB_freq, "vs. observed frequency:", observed_frequencies[2],
    "\nHWE expected BB frequency: ", exp_BB_freq, "vs. observed frequency:", observed_frequencies[3]
    )
```
For bi-allelic markers, the expected genotype frequencies under HWE are defined as $f(AA)=p^2$, $f(AB)=2pq$ and $f(BB)=q^2$ where $f(A)=p$ (frequency of allele $A$) and $f(B)=q$ (frequency of allele $B$).

If we calculate the expected genotype frequencies under HWE (as done above) and compare with observed, we see that this variant is unusual (w.r.t. HWE) in that the frequency of heterozygous alleles is $0$, which we would, in accordance with HWE, expect to be $0.49$, explaining how this variant violates the HWE significantly.

# 7. Compute the inbreeding coefficient (f) for each SNP, and make a histogram of f. You can use function HWf for this purpose. Give descriptive statistics (mean, standard deviation, etc) of f calculated over the set of SNPs. What distribution do you expect f to follow theoretically? Use a probability plot to confirm your idea.

```{r}
inbreeding_factor <- function(genotype_sequence) {
  genotype_counts <- c(
    AA=sum(genotype_sequence==0),
    AB=sum(genotype_sequence==1),
    BB=sum(genotype_sequence==2)
  )
  return(HWf(genotype_counts))
}

inbreeding_coeffs <- apply(df_mono, 2, inbreeding_factor)
hist(inbreeding_coeffs, main="Distribution of Inbreeding Coefficients for each SNP in the database", xlab="Inbreeding Coefficient")
```

```{r}
# Compute descriptive statistics
print(summary(inbreeding_coeffs))
cat("\nStandard Deviation: ", std <- sd(inbreeding_coeffs))
```
Theoretically, the distribution of inbreeding coefficients is expected to follow a normal distribution. This is because, under HWE, the genotype frequencies stabilise
```{r}
# Create probability plot with assumption of a normal distribution (qnorm)
sorted_f <- sort(inbreeding_coeffs)
empirical_prob <- seq(1, length(sorted_f)) / length(sorted_f)
qqplot(sorted_f, qnorm(empirical_prob), main = "Probability Plot", xlab = "Theoretical Quantiles", ylab = "Observed Values")
abline(0, 1, col = "red")  # Adds a 45-degree reference line for normal distribution
plot(fitted_model)
```
